### ISSUE TITLE
pl.read_excel overloads all resolve to Never, while it works correctly with `basedpyright`

### ISSUE THREAD

**Issue by shuangluoxss** [astral-sh/ty#3859]
## Description
When using ty as the language server / type checker, all overloads of polars.read_excel incorrectly resolve to Never. The same code works correctly with basedpyright, where each call resolves to the appropriate type (pl.DataFrame, dict[str, pl.DataFrame], or NoReturn).
## Reproduction
```python
import polars as pl

filename = "./1.xlsx"

df_csv = pl.read_csv(filename)
df_parquet = pl.read_parquet(filename)

res_1 = pl.read_excel(filename)
res_2 = pl.read_excel(filename, sheet_id=0)
res_3 = pl.read_excel(filename, sheet_name="Sheet1")
res_4 = pl.read_excel(filename, sheet_name=["Sheet1"])
res_5 = pl.read_excel(filename, sheet_id=0, sheet_name="Sheet1")
```
`df_csv` and `df_parquet` could be infered correctly as `pl.DataFrame`, but all `res_x` was  infered  to `Never`
## Expected Behavior
use `basedpyright` as lsp
<img width="1102" height="501" alt="Image" src="https://github.com/user-attachments/assets/4df0849e-b0a1-4056-9a5a-d03657ec34c4" />

## Real Behavior
use `ty` as lsp
<img width="1118" height="504" alt="Image" src="https://github.com/user-attachments/assets/bbae223e-dee1-40bc-aaa9-533c614ed315" />

### Version

0.0.54

**Comment by sharkdp** [comment]
Thank you for reporting this.

There is one overload of `read_excel` that returns `NoReturn` (= `Never`) [here](https://github.com/pola-rs/polars/blob/5e9cf5c09d3c0c316cba1ced8282cb835f76df5f/py-polars/src/polars/io/spreadsheet/functions.py#L159-L177). It looks like we pick this for some reason? FYI @dhruvmanila

**Comment by dhruvmanila** [comment]
A minimal reproducer:

```py
from typing import Callable, Never, ParamSpec, Protocol, TypeVar, overload, reveal_type

P = ParamSpec("P")
T = TypeVar("T")

class IdentityFunction(Protocol):
    def __call__(self, function: Callable[P, T], /) -> Callable[P, T]: ...

def identity() -> IdentityFunction:
    raise NotImplementedError

class Frame: ...

@overload
def read(value: int) -> Frame: ...
@overload
def read(value: str) -> Never: ...
@identity()
def read(value: int | str) -> Frame:
    raise NotImplementedError

reveal_type(read)
reveal_type(read(1))
```

Here, the return type of all of `read`s overload after the `identity()` decorator has been collapsed to `Never`, so I don't think it's the issue of overload matching. I think this might just be a variant of https://github.com/astral-sh/ty/issues/2799 but I thought this was fixed using a temporary workaround in https://github.com/astral-sh/ruff/pull/25806 so there are more cases remaining.

**Comment by dhruvmanila** [comment]
And, it only happens when `Never` is used as the return type to one of the overloads.

Oh, the decorator itself fails with `invalid-argument-type` which is why the workaround (https://github.com/astral-sh/ruff/pull/25806) doesn't work!

**Comment by dhruvmanila** [comment]
Going to add it as part of https://github.com/astral-sh/ty/issues/3555 to check this example.
